File: #2304 창고 다각형/구본관.py
import sys
from collections import deque
input = sys.stdin.readline

result = 0

li = []
maxh = 0

n = int(input())
for i in range(n):
    li.append(list(map(int,input().split())))
    if maxh<=li[-1][1]:
        maxh = li[-1][1]

li.sort()

preh = 0
prex = 0


for i in range(n):
    x,h = li[i]
    result += (x-prex)*preh
    prex = x
    preh = max(preh, h)
    if(h==maxh):
        break

li = li[i:]
li.sort(reverse=True)

preh = 0
prex = 1000

n = len(li)
for i in range(n):
    x,h = li[i]
    result += (prex-x)*preh
    prex = x
    preh = max(preh, h)


print(result + maxh)

File: wk02-linked-list-complete-search/#2346 풍선 터뜨리기/구본관.java
import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;
import java.util.*;

public class Main {
    public static void main(String[] args) throws IOException {
        BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
        StringBuilder sb = new StringBuilder();
        StringTokenizer st;

        int n = Integer.parseInt(br.readLine());
        int result = 0;
        int maxh = 0;

        List<int[]> li = new ArrayList<>();


        for (int i = 0; i < n; i++) {
            st = new StringTokenizer(br.readLine());
            li.add(new int[]{Integer.parseInt(st.nextToken()), Integer.parseInt(st.nextToken())});
            maxh = Math.max(maxh, li.get(i)[1]);
        }

        Collections.sort(li, new Comparator<int[]>() {
            @Override
            public int compare(int[] o1, int[] o2) {
                return o1[0]-o2[0];
            }
        });

        int preh = 0;
        int prex = 0;

        int i = 0;
        for (i = 0; i < n; i++) {
            int x = li.get(i)[0], h = li.get(i)[1];
            result += (x-prex)*preh;
            prex = x;
            preh = Math.max(preh,h);
            if(h == maxh)
                break;
        }
        
        li = li.subList(i, li.size());
        Collections.reverse(li);
        
        preh = 0;
        prex = 1000;
        
        n = li.size();
        for (i = 0; i < n; i++) {
        	int x = li.get(i)[0], h = li.get(i)[1];
        	result += (prex-x)*preh;
        	prex = x;
        	preh = Math.max(preh,h);
        }
        
        sb.append(result + maxh);


        System.out.println(sb);
    }
}
